### ISSUE TITLE
trackBy is broken in version 9.3

### ISSUE THREAD

**Issue by robleka** [tiberiuzuld/angular-gridster2#645]
Hello. 

I'm using gridster as a dashboard creating tool, so items dynamically change (they can be added or removed on demand). To prevent refreshing item views too many times, I'm using my own trackBy function. Up to 9.2 version everything was working as it should, but now i get the following errors when adding and removing the same item in the same "session" (without reloading page / gridster in between):

`vpsc.ts:554 ERROR TypeError: Cannot read property 'gridRenderer' of undefined
    at GridsterItemComponent.push../node_modules/angular-gridster2/__ivy_ngcc__/fesm5/angular-gridster2.js.GridsterItemComponent.setSize (vpsc.ts:554)
    at GridsterComponent.push../node_modules/angular-gridster2/__ivy_ngcc__/fesm5/angular-gridster2.js.GridsterComponent.calculateLayout (vpsc.ts:554)
    at later (vpsc.ts:554)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.js:3504)
    at Object.onInvokeTask (vpsc.ts:554)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.js:3503)
    at Zone.push../node_modules/zone.js/dist/zone.js.Zone.runTask (polyfills.js:3271)
    at push../node_modules/zone.js/dist/zone.js.ZoneTask.invokeTask (polyfills.js:3586)
    at ZoneTask.invoke (polyfills.js:3575)
    at timer (polyfills.js:6117)`

I tried to change my trackBy function to only track by index and the problem remains.

I saw that some additional trackBy functionalities were added in 9.3. Would it be possible, to add option to disable them on demand?

Thanks.

**Comment by tiberiuzuld** [comment]
Hi @robleka ,
I was not able to reproduce your issue using the steps described by you.
I improved the demo for track by in an attempt to reproduce the error you are encountering.
https://tiberiuzuld.github.io/angular-gridster2/trackBy

The changes made to better support trackBy was just a change from ngOnInit to ngOnChanges so the components use to render each item are not recreated and instead reused.
You can find the changes here https://github.com/tiberiuzuld/angular-gridster2/pull/630/files

Instead of adding another option, I would prefer to fix the issue you are encountering.

Regarding the log you added seems like gridsterItem was destroyed and the `setSize` was being called:
```
  setSize(): void {
    this.renderer.setStyle(this.el, 'display', this.notPlaced ? '' : 'block');
    this.gridster.gridRenderer.updateItem(this.el, this.$item, this.renderer);
    this.updateItemSize();
  }
   ngOnDestroy(): void {
    this.gridster.removeItem(this);
    delete this.gridster;
    this.drag.destroy();
    delete this.drag;
    this.resize.destroy();
    delete this.resize;
  }
  constructor(@Inject(ElementRef) el: ElementRef, gridster: GridsterComponent, @Inject(Renderer2) public renderer: Renderer2,
              @Inject(NgZone) private zone: NgZone) {
    this.el = el.nativeElement;
    this.$item = {
      cols: -1,
      rows: -1,
      x: -1,
      y: -1,
    };
    this.gridster = gridster;
    this.drag = new GridsterDraggable(this, gridster, this.zone);
    this.resize = new GridsterResizable(this, gridster, this.zone);
  }
```
This maybe be possible if the `calculateLayout` was deferred and the item was removed while doing that.

**Comment by robleka** [comment]
Hello,

Thanks for response. I think I may have solved my problem. My application works like this: 
1. User removes or changes (position, size) gridster item or content of the item (chart configuration for example)
2. User confirms changes which are then send to server
3. Server returns complete dashboard (all items, not only modified)
4. I process the dashboard and prepare gridster items
5. I replace the whole gridster items array at once (trackBy indexes for not-modified items stay the same as before)

I now added a timeout between steps 4 and 5 and error went away. It looks to me like something still needs to access old items right at the moment when I'm changing the array. With timeout added that operation can finish before items are replaced with new ones. 

So I'm closing this issue and will reopen it if problem returns.

**Comment by tmtron** [comment]
I think I hit the same problem.

IMHO: adding a timeout is not a solution - not even a workaround, because then it may sometimes work and sometimes fail: e.g. when the CPU is busy or on slow mobile devices, etc.

Here is a [Stackblitz example](https://stackblitz.com/edit/angular-gridster-track-by3?file=src/app/app.component.ts) to reproduce the issue.  

Notes:  
- the example uses the default change detection strategy (**not** `OnPush`)
- the reset feature uses the workaround as explained in #644 

Reproduction steps:  
- click on the `ID: 0` text to remove the item
- click reset
- click on the `ID: 0` again
- check the errors in the console (at the bottom of the stackblitz browser section

It seems that the error occurs in the `setSize` function as outlined in the comment above.

>This maybe be possible if the calculateLayout was deferred and the item was removed while doing that.

In this case, maybe the `calculateLayout` function can first check if this item has already been destroyed?

**Comment by tiberiuzuld** [comment]
Hi @tmtron ,
I think you are right.
I will have to add a check to see if the item is destroyed before trying to calculate the setSize.
I will try make a fix soon.

Thanks

**Comment by tiberiuzuld** [comment]
So after the investigation, seems like the issue was not with the destroy itself but with the init.
On the reset seems like the item was added multiple times to the internal list of items because the `init` property was set to false only after the debounced calculation of the layout, and in that time the item was added 2 times when reseting with trackBy.
When removing the item the destroy was called 2 times and so the error appeared.
Fixed the issue by adding the item only on ngOnInit instead of the condition inside the ngOnChanges

**Comment by tiberiuzuld** [comment]
Fixed in v10.1.3
Thank you for reporting the issue and reproducing in stackblits
@tmtron and @robleka

**Comment by tmtron** [comment]
@tiberiuzuld that was  a fast fix. Thanks a lot for your hard work on this lib!
